app/Models/Concerns/HasMemberships.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+namespace App\Models\Concerns;
+
+use App\Exceptions\Memberships\MembershipNotSecondaryException;
+use App\Exceptions\Memberships\PrimaryMembershipDoesntAllowSecondaryException;
+use App\Models\Membership;
+use App\Models\Membership\MembershipPivot;
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Relations\BelongsToMany;
+
+trait HasMemberships
+{
+    private $primaryMembership;
+
+    public function memberships(): BelongsToMany
+    {
+        return $this->membershipsRelationship()
+            ->wherePivot('ended_at', null)
+            ->orderByPriority();
+    }
+
+    public function membershipHistory(): BelongsToMany
+    {
+        return $this->membershipsRelationship()
+            ->orderBy((new MembershipPivot())->getTable().'.started_at', 'desc');
+    }
+
+    public function primaryMembership(): ?Membership
+    {
+        if ($this->primaryMembership) {
```

Comment:
What is the benefit of this if check? I can see it is stored as a private attribute on this trait but why?

Reply:
Because getting the primary membership isn't an eloquent relationship, it won't be cached (i.e. a database call is made with every call to the function). By doing this, we replicate eloquent behaviour of it caching the response on initial call